rust/src/normalization.rs

```diff
@@ -120,6 +121,294 @@ pub(crate) fn rewrite_rpmdb_timestamps<F: Read + Write + Seek>(rpmdb: &mut F) ->
     Ok(())
 }
 
+#[context("Rewriting rpmdb database files for build stability")]
+pub(crate) fn normalize_rpmdb(rootfs: &openat::Dir, rpmdb_path: impl AsRef<Path>) -> Result<()> {
+    let source_date = if let Some(source_date) = source_date_epoch() {
+        source_date as u32
+    } else {
+        return Ok(());
+    };
+
+    let mut db_backend = None;
+
+    let macros = BufReader::new(rootfs.open_file("usr/lib/rpm/macros")?);
+    for line in macros.lines() {
+        let line = line?;
+        if line.is_empty() {
+            continue;
+        }
+
+        let mut bits = line.split(char::is_whitespace).filter(|s| !s.is_empty());
+        if let Some(m) = bits.next() {
+            if m == "%_db_backend" {
+                db_backend = bits.last().map(|s| s.to_string());
+                break;
+            }
+        };
+    }
+
+    let db_backend = db_backend.ok_or_else(|| anyhow!("Unable to determine rpmdb backend"))?;
+
+    match db_backend.as_str() {
+        "bdb" | "b-d-b" => rpmdb_bdb::normalize_rpmdb_bdb(rootfs, rpmdb_path, source_date),
+        "ndb" => Ok(()),
+        "sqlite" => Ok(()),
+        _ => Err(anyhow!("Unknown rpmdb backend: {}", db_backend)),
+    }
+}
+
+mod rpmdb_bdb {
+    use anyhow::{anyhow, Context, Result};
+    use binread::{BinRead, BinReaderExt};
+    use lazy_static::lazy_static;
+    use openat::SimpleType;
+    use sha2::Digest;
+    use std::ffi::OsStr;
+    use std::io::{Read, Seek, SeekFrom, Write};
+    use std::os::unix::io::AsRawFd;
+    use std::path::{Path, PathBuf};
+    use subprocess::{Exec, Redirection};
+
+    #[derive(BinRead, Debug, Clone, Copy, PartialEq, Eq)]
+    #[repr(u8)]
+    #[br(repr=u8)]
+    enum PageType {
+        IBTree = 3,
+        LBTree = 5,
+        Overflow = 7,
+        HashMeta = 8,
+        BTreeMeta = 9,
+        Hash = 13,
+    }
+
+    #[derive(BinRead, Debug)]
+    #[br(little)]
+    struct MetaHeader {
+        lsn: u64,
+        pgno: u32,
+        magic: u32,
+        version: u32,
+        pagesize: u32,
+        encrypt_alg: u8,
+        page_type: PageType,
+        metaflags: u8,
+        unused1: u8,
+        free: u32,
+        last_pgno: u32,
+        nparts: u32,
+        key_count: u32,
+        record_count: u32,
+        flags: u32,
+        uid: [u8; 20],
+    }
+
+    const BTREE_MAGIC: u32 = 0x00053162;
+    const HASH_MAGIC: u32 = 0x00061561;
+    const PAGE_HEADER_SIZE: u16 = 26;
+    const PAGE_HEADER_MAGIC_OFFSET: u64 = 0x34;
+
+    #[derive(BinRead, Debug)]
+    #[br(little)]
+    struct PageHeader {
```

Comment:
I think it's worth a comment here around how you created these structures.

Did you just...read the source code to bdb and type it in?

Reply:
Pretty much, yep.